File: src/openclaw_quant/indicators.py
"""
Technical Indicators Library

Provides common technical analysis indicators using pandas/numpy.
Optimized for vectorized operations.
"""

import numpy as np
import pandas as pd
from typing import Union


def SMA(series: Union[pd.Series, np.ndarray], period: int) -> pd.Series:
    """Simple Moving Average"""
    if isinstance(series, np.ndarray):
        series = pd.Series(series)
    return series.rolling(window=period).mean()


def EMA(series: Union[pd.Series, np.ndarray], period: int) -> pd.Series:
    """Exponential Moving Average"""
    if isinstance(series, np.ndarray):
        series = pd.Series(series)
    return series.ewm(span=period, adjust=False).mean()


def RSI(series: Union[pd.Series, np.ndarray], period: int = 14) -> pd.Series:
    """
    Relative Strength Index

    Args:
        series: Price series
        period: RSI period (default: 14)

    Returns:
        RSI values (0-100)
    """
    if isinstance(series, np.ndarray):
        series = pd.Series(series)

    delta = series.diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()

    rs = gain / loss
    rsi = 100 - (100 / (1 + rs))

    return rsi


def MACD(series: Union[pd.Series, np.ndarray],
         fast: int = 12, slow: int = 26, signal: int = 9) -> pd.DataFrame:
    """
    Moving Average Convergence Divergence

    Returns:
        DataFrame with 'macd', 'signal', and 'histogram' columns
    """
    if isinstance(series, np.ndarray):
        series = pd.Series(series)

    ema_fast = EMA(series, fast)
    ema_slow = EMA(series, slow)

    macd_line = ema_fast - ema_slow
    signal_line = EMA(macd_line, signal)
    histogram = macd_line - signal_line

    return pd.DataFrame({
        'macd': macd_line,
        'signal': signal_line,
        'histogram': histogram
    })


def BollingerBands(series: Union[pd.Series, np.ndarray],
                   period: int = 20, std_dev: float = 2.0) -> pd.DataFrame:
    """
    Bollinger Bands

    Returns:
        DataFrame with 'upper', 'middle', and 'lower' columns
    """
    if isinstance(series, np.ndarray):
        series = pd.Series(series)

    middle = SMA(series, period)
    std = series.rolling(window=period).std()

    upper = middle + (std * std_dev)
    lower = middle - (std * std_dev)

    return pd.DataFrame({
        'upper': upper,
        'middle': middle,
        'lower': lower
    })


def ATR(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14) -> pd.Series:
    """
    Average True Range

    Args:
        high: High prices
        low: Low prices
        close: Close prices
        period: ATR period

    Returns:
        ATR values
    """
    tr1 = high - low
    tr2 = abs(high - close.shift())
    tr3 = abs(low - close.shift())

    tr = pd.concat([tr1, tr2, tr3], axis=1).max(axis=1)
    atr = tr.rolling(window=period).mean()

    return atr


def Stochastic(high: pd.Series, low: pd.Series, close: pd.Series,
               period: int = 14, smooth_k: int = 3, smooth_d: int = 3) -> pd.DataFrame:
    """
    Stochastic Oscillator

    Returns:
        DataFrame with '%K' and '%D' columns
    """
    lowest_low = low.rolling(window=period).min()
    highest_high = high.rolling(window=period).max()

    k_raw = 100 * (close - lowest_low) / (highest_high - lowest_low)
    k = k_raw.rolling(window=smooth_k).mean()
    d = k.rolling(window=smooth_d).mean()

    return pd.DataFrame({
        '%K': k,
        '%D': d
    })


def ADX(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14) -> pd.Series:
    """Average Directional Index"""
    plus_dm = high.diff()
    minus_dm = -low.diff()

    plus_dm[plus_dm < 0] = 0
    minus_dm[minus_dm < 0] = 0

    tr = ATR(high, low, close, 1)

    plus_di = 100 * (plus_dm.rolling(window=period).mean() / tr.rolling(window=period).mean())
    minus_di = 100 * (minus_dm.rolling(window=period).mean() / tr.rolling(window=period).mean())

    dx = 100 * abs(plus_di - minus_di) / (plus_di + minus_di)
    adx = dx.rolling(window=period).mean()

    return adx


def OBV(close: pd.Series, volume: pd.Series) -> pd.Series:
    """On-Balance Volume"""
    obv = (np.sign(close.diff()) * volume).fillna(0).cumsum()
    return obv


def VWAP(high: pd.Series, low: pd.Series, close: pd.Series, volume: pd.Series) -> pd.Series:
    """Volume Weighted Average Price"""
    typical_price = (high + low + close) / 3
    vwap = (typical_price * volume).cumsum() / volume.cumsum()
    return vwap


def ROC(series: Union[pd.Series, np.ndarray], period: int = 12) -> pd.Series:
    """Rate of Change"""
    if isinstance(series, np.ndarray):
        series = pd.Series(series)

    roc = ((series - series.shift(period)) / series.shift(period)) * 100
    return roc


def CCI(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 20) -> pd.Series:
    """Commodity Channel Index"""
    typical_price = (high + low + close) / 3
    sma_tp = typical_price.rolling(window=period).mean()
    mean_deviation = typical_price.rolling(window=period).apply(
        lambda x: np.abs(x - x.mean()).mean()
    )

    cci = (typical_price - sma_tp) / (0.015 * mean_deviation)
    return cci


def WilliamsR(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14) -> pd.Series:
    """Williams %R"""
    highest_high = high.rolling(window=period).max()
    lowest_low = low.rolling(window=period).min()

    williams_r = -100 * (highest_high - close) / (highest_high - lowest_low)
    return williams_r


# Convenience functions for strategy use
def crossover(series1: pd.Series, series2: pd.Series) -> pd.Series:
    """Detect crossover (series1 crosses above series2)"""
    return (series1 > series2) & (series1.shift(1) <= series2.shift(1))


def crossunder(series1: pd.Series, series2: pd.Series) -> pd.Series:
    """Detect crossunder (series1 crosses below series2)"""
    return (series1 < series2) & (series1.shift(1) >= series2.shift(1))


__all__ = [
    'SMA', 'EMA', 'RSI', 'MACD', 'BollingerBands',
    'ATR', 'Stochastic', 'ADX', 'OBV', 'VWAP',
    'ROC', 'CCI', 'WilliamsR',
    'crossover', 'crossunder'
]

File: src/utils/indicators.ts
/**
 * Technical Indicators Library
 * Common technical analysis indicators for trading strategies
 */

import { Candle } from '../types';

export class Indicators {
  /**
   * Simple Moving Average (SMA)
   */
  static sma(values: number[], period: number): number[] {
    const result: number[] = [];

    for (let i = 0; i < values.length; i++) {
      if (i < period - 1) {
        result.push(NaN);
        continue;
      }

      const sum = values.slice(i - period + 1, i + 1).reduce((a, b) => a + b, 0);
      result.push(sum / period);
    }

    return result;
  }

  /**
   * Exponential Moving Average (EMA)
   */
  static ema(values: number[], period: number): number[] {
    const result: number[] = [];
    const multiplier = 2 / (period + 1);

    let ema = values.slice(0, period).reduce((a, b) => a + b, 0) / period;

    for (let i = 0; i < values.length; i++) {
      if (i < period - 1) {
        result.push(NaN);
      } else if (i === period - 1) {
        result.push(ema);
      } else {
        ema = (values[i] - ema) * multiplier + ema;
        result.push(ema);
      }
    }

    return result;
  }

  /**
   * Relative Strength Index (RSI)
   */
  static rsi(values: number[], period: number = 14): number[] {
    const result: number[] = [];
    const gains: number[] = [];
    const losses: number[] = [];

    for (let i = 1; i < values.length; i++) {
      const change = values[i] - values[i - 1];
      gains.push(change > 0 ? change : 0);
      losses.push(change < 0 ? -change : 0);
    }

    const avgGains = this.ema(gains, period);
    const avgLosses = this.ema(losses, period);

    result.push(NaN); // First value is always NaN

    for (let i = 0; i < avgGains.length; i++) {
      if (isNaN(avgGains[i]) || isNaN(avgLosses[i])) {
        result.push(NaN);
        continue;
      }

      const rs = avgGains[i] / avgLosses[i];
      const rsi = 100 - (100 / (1 + rs));
      result.push(rsi);
    }

    return result;
  }

  /**
   * Bollinger Bands
   */
  static bollingerBands(
    values: number[],
    period: number = 20,
    stdDev: number = 2
  ): { upper: number[]; middle: number[]; lower: number[] } {
    const middle = this.sma(values, period);
    const upper: number[] = [];
    const lower: number[] = [];

    for (let i = 0; i < values.length; i++) {
      if (i < period - 1) {
        upper.push(NaN);
        lower.push(NaN);
        continue;
      }

      const slice = values.slice(i - period + 1, i + 1);
      const mean = middle[i];
      const variance = slice.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / period;
      const std = Math.sqrt(variance);

      upper.push(mean + std * stdDev);
      lower.push(mean - std * stdDev);
    }

    return { upper, middle, lower };
  }

  /**
   * Moving Average Convergence Divergence (MACD)
   */
  static macd(
    values: number[],
    fastPeriod: number = 12,
    slowPeriod: number = 26,
    signalPeriod: number = 9
  ): { macd: number[]; signal: number[]; histogram: number[] } {
    const fastEma = this.ema(values, fastPeriod);
    const slowEma = this.ema(values, slowPeriod);

    const macd = fastEma.map((fast, i) => fast - slowEma[i]);
    const signal = this.ema(macd.filter(v => !isNaN(v)), signalPeriod);

    // Pad signal array to match length
    const paddedSignal = new Array(macd.length - signal.length).fill(NaN).concat(signal);
    const histogram = macd.map((m, i) => m - paddedSignal[i]);

    return { macd, signal: paddedSignal, histogram };
  }

  /**
   * Average True Range (ATR)
   */
  static atr(candles: Candle[], period: number = 14): number[] {
    const trueRanges: number[] = [];

    for (let i = 1; i < candles.length; i++) {
      const high = candles[i].high;
      const low = candles[i].low;
      const prevClose = candles[i - 1].close;

      const tr = Math.max(
        high - low,
        Math.abs(high - prevClose),
        Math.abs(low - prevClose)
      );

      trueRanges.push(tr);
    }

    const atr = this.ema(trueRanges, period);
    return [NaN, ...atr]; // Add NaN for first candle
  }

  /**
   * Stochastic Oscillator
   */
  static stochastic(
    candles: Candle[],
    period: number = 14,
    kSmooth: number = 3,
    dSmooth: number = 3
  ): { k: number[]; d: number[] } {
    const k: number[] = [];

    for (let i = 0; i < candles.length; i++) {
      if (i < period - 1) {
        k.push(NaN);
        continue;
      }

      const slice = candles.slice(i - period + 1, i + 1);
      const high = Math.max(...slice.map(c => c.high));
      const low = Math.min(...slice.map(c => c.low));
      const close = candles[i].close;

      const stoch = ((close - low) / (high - low)) * 100;
      k.push(stoch);
    }

    const kSmoothed = this.sma(k.filter(v => !isNaN(v)), kSmooth);
    const d = this.sma(kSmoothed.filter(v => !isNaN(v)), dSmooth);

    // Pad arrays
    const paddedK = new Array(candles.length - kSmoothed.length).fill(NaN).concat(kSmoothed);
    const paddedD = new Array(candles.length - d.length).fill(NaN).concat(d);

    return { k: paddedK, d: paddedD };
  }

  /**
   * Volume Weighted Average Price (VWAP)
   */
  static vwap(candles: Candle[]): number[] {
    const result: number[] = [];
    let cumulativeTPV = 0;
    let cumulativeVolume = 0;

    for (const candle of candles) {
      const typicalPrice = (candle.high + candle.low + candle.close) / 3;
      cumulativeTPV += typicalPrice * candle.volume;
      cumulativeVolume += candle.volume;

      result.push(cumulativeTPV / cumulativeVolume);
    }

    return result;
  }

  /**
   * On-Balance Volume (OBV)
   */
  static obv(candles: Candle[]): number[] {
    const result: number[] = [0];

    for (let i = 1; i < candles.length; i++) {
      const prev = result[i - 1];
      const change = candles[i].close > candles[i - 1].close ? candles[i].volume :
                    candles[i].close < candles[i - 1].close ? -candles[i].volume : 0;
      result.push(prev + change);
    }

    return result;
  }

  /**
   * Helper: Get closes from candles
   */
  static getCloses(candles: Candle[]): number[] {
    return candles.map(c => c.close);
  }

  /**
   * Helper: Get highs from candles
   */
  static getHighs(candles: Candle[]): number[] {
    return candles.map(c => c.high);
  }

  /**
   * Helper: Get lows from candles
   */
  static getLows(candles: Candle[]): number[] {
    return candles.map(c => c.low);
  }

  /**
   * Helper: Get volumes from candles
   */
  static getVolumes(candles: Candle[]): number[] {
    return candles.map(c => c.volume);
  }
}
